src/datum_api.c

```diff
@@ -1568,6 +1568,67 @@ int datum_api_OK(struct MHD_Connection *connection) {
 	return datum_api_submit_uncached_response(connection, MHD_HTTP_OK, response);
 }
 
+int datum_api_umbrel_widget(struct MHD_Connection * const connection) {
+    char json_response[512];
+	int j, k = 0, ii;
+	T_DATUM_MINER_DATA *m;
+	T_DATUM_API_DASH_VARS umbreldata;
+	unsigned char astat;
+	double thr = 0.0;
+	double hr;
+	uint64_t tsms;
+
+	memset(&umbreldata, 0, sizeof(T_DATUM_API_DASH_VARS));
+
+	pthread_rwlock_rdlock(&stratum_global_job_ptr_lock);
+	j = global_latest_stratum_job_index;
+	umbreldata.sjob = (j >= 0 && j < MAX_STRATUM_JOBS) ? global_cur_stratum_jobs[j] : NULL;
+	pthread_rwlock_unlock(&stratum_global_job_ptr_lock);
+
+	tsms = current_time_millis();
+
+	if (global_stratum_app) {
+		k = 0;
+		for(j=0;j<global_stratum_app->max_threads;j++) {
+			k+=global_stratum_app->datum_threads[j].connected_clients;
+			for(ii=0;ii<global_stratum_app->max_clients_thread;ii++) {
+				if (global_stratum_app->datum_threads[j].client_data[ii].fd > 0) {
+					m = (T_DATUM_MINER_DATA *)global_stratum_app->datum_threads[j].client_data[ii].app_client_data;
+					if (m->subscribed) {
+						astat = m->stats.active_index?0:1; // inverted
+						hr = 0.0;
+						if ((m->stats.last_swap_ms > 0) && (m->stats.diff_accepted[astat] > 0)) {
+							hr = ((double)m->stats.diff_accepted[astat] / (double)((double)m->stats.last_swap_ms/1000.0)) * 0.004294967296;
+						}
+						if (((double)(tsms - m->stats.last_swap_tsms)/1000.0) < 180.0) {
+							thr += hr;
+						}
+					}
+				}
+			}
+		}
+		umbreldata.STRATUM_TOTAL_CONNECTIONS = k;
+		umbreldata.STRATUM_HASHRATE_ESTIMATE = thr;
+	} else {
+		umbreldata.STRATUM_TOTAL_CONNECTIONS = 0;
+		umbreldata.STRATUM_HASHRATE_ESTIMATE = 0.0;
+	}
+
+    snprintf(json_response, sizeof(json_response), "{"
+        "\"type\": \"two-stats\","
+        "\"refresh\": \"30s\","
+        "\"link\": \"\","
+        "\"items\": ["
+        "    {\"title\": \"Connections\", \"text\": %d, \"subtext\": \"Worker\"},"
+        "    {\"title\": \"Hashrate\", \"text\": %.2f, \"subtext\": \"TH/s\"}"
+        "]"
+    "}", umbreldata.STRATUM_TOTAL_CONNECTIONS, umbreldata.STRATUM_HASHRATE_ESTIMATE);
+
+    struct MHD_Response *response = MHD_create_response_from_buffer(strlen(json_response), (void *)json_response, MHD_RESPMEM_PERSISTENT);
```

Comment:
Since you're using `snprintf` right above, save its return value and use `json_response_len < sizeof(json_response) ? json_response_len : (sizeof(json_response) - 1)`